### ISSUE TITLE
Oracle JDK 9 not installed despite JDK oraclejdk9

### ISSUE THREAD

**Issue by wltjr** [Shippable/support#4156]
### Description of your issue:
This has been driving me crazy for hours. I finally know what is going on. My build shows.
```
JDK oraclejdk9
```
But all I see is anything but, and ancient 1.6 stuff... I see alternatives setting to oracle 7...
```
setup_jdk
sudo update-alternatives --set java /usr/lib/jvm/java-7-oracle/jre/bin/java
sudo update-alternatives --set javac /usr/lib/jvm/java-7-oracle/bin/javac
```
```
 ls -l /usr/lib/jvm
total 16
lrwxrwxrwx 1 root root   24 Mar 22  2014 default-java -> java-1.7.0-openjdk-amd64
lrwxrwxrwx 1 root root   20 Jan 26  2016 java-1.6.0-openjdk-amd64 -> java-6-openjdk-amd64
lrwxrwxrwx 1 root root   20 Jan 21  2016 java-1.7.0-openjdk-amd64 -> java-7-openjdk-amd64
drwxr-xr-x 7 root root 4096 Feb 15  2016 java-6-openjdk-amd64
drwxr-xr-x 3 root root 4096 Feb 15  2016 java-6-openjdk-common
drwxr-xr-x 7 root root 4096 Feb 15  2016 java-7-openjdk-amd64
drwxr-xr-x 8 root root 4096 Feb 15  2016 java-7-oracle
```

https://app.shippable.com/github/Obsidian-StudiosInc/jem/runs/108/1/console

Its been failing since I stopped using openjdk 7
https://app.shippable.com/github/Obsidian-StudiosInc/jem/runs/98/1/console
https://app.shippable.com/github/Obsidian-StudiosInc/jem/dashboard

I have tried using different nodes, I updated the version and tried using unstable. I am on Ubuntu 14.04, not sure if that is the problem.

**Comment by rageshkrishna** [comment]
@wltjr Your build is currently running on the  `drydock/u16cppall:v6.1.4` image. The CPP images do not have multiple versions of Java available as they were intended specifically for cpp toolchains. You should try using `drydock/u16javall:v6.1.4` instead if you need other versions of Java. 

We've documented what's available in this image here: http://docs.shippable.com/platform/runtime/machine-image/ami-v614/#jav-614. Hopefully the java image has everything you need to run your builds.

**Comment by wltjr** [comment]
I mostly build C projects and this is a C project not Java. Why is it saying JDK oraclejdk9 if that is not the case? Why are the others present at all? I haven't any of these problems on Travis. Not sure why i need to deal with the images and nodes. Seems like more work...

**Comment by rageshkrishna** [comment]
What I said was that the CPP images do not have multiple versions of Java available. They do have _a_ Java version, which is currently 1.7.

Your [YML asks for oraclejdk9](https://github.com/Obsidian-StudiosInc/jem/blob/master/.travis.yml#L9), so we display that value as what the job is configured to do. This is also why we attempt to set a Java version during the build at all.

I hope this makes things clearer.

**Comment by wltjr** [comment]
If you see the output above there is 1 Java 6 and 2 Java7s, openjdk and oraclejdk. Just not 8 or not. You keep saying it does not have multiple versions and it clearly does. Why before I switched to 9 things worked. I assumed since it showed JDK 9 in the build information. That was a real setting being used, not some fake value with the actual system being other than what I am after.

How can I modify things then to get Java 9? I guess I may need to install that package or do something specific for Shippable. I rather not have to go back to Java 7 because that is all that is on some current image. When Oracle 1.7 is EOL. Surprised 6 is there at all, much less 7 and total lack of 8 or 9. Again none would make sense. but there are 3 present now, all old. Despite you saying it does not have multiple versions.

I have gone out of my way on most all projects to make a .travis.yml that is parse-able by shippable. It is to much time to maintain 2, and given various issues. I tend to prefer Travis and use Shippable more as a fallback. These kinds of issues seem more like limitations and reasons to not use Shippable.

How do I proceed here?

**Comment by wltjr** [comment]
I guess for now I will need to revert back to openjdk-7. I have at least modified things so I can switch back and forth. I am not building anything Java. This is a Java Environment Manager, like jdk_switcher or alternatives, but with many more features. Similar but different.

**Comment by rageshkrishna** [comment]
@wltjr I went back and looked at the source to double-check this. You're right, this image does have multiple JDKs. Because of the way the images inherit from one another, I missed the fact that additional Java's were available in the CPP image. I stand corrected on this.

As I see it, there are two things you can try:

1. You can install JDK 9 during the build. You can take a look at [the script we use](https://github.com/dry-dock/u14javall/blob/master/version/oraclejdk9.sh) to install it in the u14javall image as a reference.
1. You could try using the u16javall image itself, but this would require using a shippable.yml instead.

In the mean time, I'll also look in to making JDK 9 available on all the images in the next drydock release.

**Comment by wltjr** [comment]
@rageshkrishna ok thanks! No worries on the multiple jdk just was differing from what I was seeing. I myself am still behind on all the docker stuff. I need to catch up there. I have a single android project that I never got building under Shippable I could likely address via similar means.

Thanks for the link, I will look into that, and/or switching to openjdk-7 for shippable. I made changes such that the vm used for testing is provided via variables. Not hard to swap those out in the [travis.yml](https://github.com/Obsidian-StudiosInc/jem/blob/master/.travis.yml#L74). For [some others](https://github.com/Obsidian-StudiosInc/asspr/blob/master/.travis.yml#L20) I do that based on ```SHIPPABLE``` env variable. I can look to do something similar there.

For images in general, you may consider providing [jem](https://github.com/Obsidian-StudiosInc/jem) someday as well. It basically does what jdk_switcher does and a lot more.

Anyway thanks much for offering to address the issue in a future release!

**Comment by wltjr** [comment]
Not sure if you need bother with Java 9, seems that can be skipped for Java 10. May need to prep for Java 11 as well. For time being I have switched to a method that works for both travis and shippable, based on your [oraclejdk10.sh](https://github.com/dry-dock/u14javall/blob/master/version/oraclejdk10.sh). I made those same changes to my [.travis.yml](https://github.com/Obsidian-StudiosInc/jem/commit/08b3ae98aea118d0ce6f437e3c7d910fd8f013aa), and now have it consistent for both. Well less the [one line](https://github.com/dry-dock/u14javall/blob/master/version/oraclejdk10.sh#L4) that seems specific to Shippable. For what ever reason I do not have to do that on Travis.

Thanks again for those files!!! I went straight there this time. I noticed the same for a [Java 10 issue for Travis](https://github.com/travis-ci/travis-ci/issues/9368#issuecomment-383980483). You can see it does not have that  [one line](https://github.com/dry-dock/u14javall/blob/master/version/oraclejdk10.sh#L4) that is unique to Shippable. That is minor and moot. Anyway thanks again!

**Comment by manishas** [comment]
@rageshkrishna can you post a resolution for this issue?

**Comment by rageshkrishna** [comment]
I just checked in with the Drydock team. All images will contain OpenJDK 10 and OracleJDK 10 in the v6.8.4 release.

**Comment by wltjr** [comment]
FYI 11 will be released officially in [~2 months](http://www.java-countdown.xyz/), and is a [LTS major version](https://cdn-images-1.medium.com/max/1200/1*OoQAbZ28JQoy-B98nvgXqg.png) for at least 3yrs. May want to leave this open till then, unless another exists.

**Comment by rageshkrishna** [comment]
JDK 11 is going to be available in the v6.10.4 images coming out next week.

**Comment by nealeu** [comment]
v6.10.4 is now out.  @rageshkrishna can you confirm that this covers JDK 11 on the CPP builds too?

Scanning the above comments, it would make sense to have JDK8 and 11. 

As far as I can tell the only supported versions of Java are currently 8, 11 and 12.  And from January, 8 will not be officially supported by Oracle.

**Comment by rageshkrishna** [comment]
@nealeu Yes, v6.10.4 has Java 11 in the CPP images (both openjdk and oraclejdk). In an effort to keep image sizes low, the CPP images now have only Java 11 on them.

Other versions of Java are available in the Java images.

**Comment by nealeu** [comment]
@rageshkrishna Thanks for clarifying.  Sounds a sane move.

**Comment by rageshkrishna** [comment]
I'm closing this as we now have images with Java 11.
